xnmt/decoder.py

```diff
@@ -152,3 +161,120 @@ def calc_loss(self, mlp_dec_state, ref_action):
+               mlp_layer=bare(MLP),
+               bridge=bare(CopyBridge),
+               label_smoothing=0.0,
+               lexicon_file=None,
+               src_vocab=Ref(Path("model.src_reader.vocab")),
+               trg_vocab=Ref(Path("model.trg_reader.vocab")),
+               attender=Ref(Path("model.attender")),
+               lexicon_type='bias',
+               lexicon_alpha=0.001,
+               linear_projector=None,
+               param_init_lin=Ref("exp_global.param_init", default=bare(GlorotInitializer)),
+               bias_init_lin=Ref("exp_global.bias_init", default=bare(ZeroInitializer)),
+               ):
+    super().__init__(input_dim, trg_embed_dim, input_feeding, rnn_layer,
+                     mlp_layer, bridge, label_smoothing)
+    assert lexicon_file is not None
+    self.lexicon_file = lexicon_file
+    self.src_vocab = src_vocab
+    self.trg_vocab = trg_vocab
+    self.attender = attender
+    self.lexicon_type = lexicon_type
+    self.lexicon_alpha = lexicon_alpha
+
+    self.linear_projector = self.add_serializable_component("linear_projector", linear_projector,
+                                                             lambda: xnmt.linear.Linear(input_dim=input_dim,
+                                                                                        output_dim=mlp_layer.output_dim))
+
+    if self.lexicon_type == "linear":
+      self.lexicon_method = self.linear
+    elif self.lexicon_type == "bias":
+      self.lexicon_method = self.bias
+    else:
+      raise ValueError("Unrecognized lexicon method:", lexicon_type, "can only choose between [bias, linear]")
+
+  def load_lexicon(self):
+    logger.info("Loading lexicon from file: " + self.lexicon_file)
+    assert self.src_vocab.frozen
+    assert self.trg_vocab.frozen
+    lexicon = [{} for _ in range(len(self.src_vocab))]
+    with open(self.lexicon_file) as fp:
+      for line in fp:
+        try:
+          trg, src, prob = line.rstrip().split()
+        except:
+          logger.warning("Failed to parse 'trg src prob' from:" + line.strip())
+          continue
+        trg_id = self.trg_vocab.convert(trg)
+        src_id = self.src_vocab.convert(src)
+        lexicon[src_id][trg_id] = float(prob)
+    # Setting the rest of the weight to the unknown word
+    for i in range(len(lexicon)):
+      sum_prob = sum(lexicon[i].values())
+      if sum_prob < 1.0:
+        lexicon[i][self.trg_vocab.convert(self.trg_vocab.unk_token)] = 1.0 - sum_prob
+    # Overriding special tokens
+    src_unk_id = self.src_vocab.convert(self.src_vocab.unk_token)
+    trg_unk_id = self.trg_vocab.convert(self.trg_vocab.unk_token)
+    lexicon[self.src_vocab.SS] = {self.trg_vocab.SS: 1.0}
+    lexicon[self.src_vocab.ES] = {self.trg_vocab.ES: 1.0}
+    # TODO(philip30): Note sure if this is intended
+    lexicon[src_unk_id] = {trg_unk_id: 1.0}
+    return lexicon
+
+  @handle_xnmt_event
+  def on_new_epoch(self, training_task, *args, **kwargs):
```

Comment:
Why is this in `on_new_epoch` and not in the constructor?

Reply:
I was afraid that the vocabulary has not been populated in case of vocab is constructed when reading the corpus. In this case, the length of the vocabulary will be zero. on_new_epoch will ensure that the vocabulary has been populated so we get the correct length

Reply:
Ahh, fair enough.